src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/graphs/GraphBasedKBestHaplotypeFinder.java

```diff
@@ -0,0 +1,154 @@
+                    final Set<E> outgoingEdges = graph.outgoingEdgesOf(vertexToExtend);
+                    int totalOutgoingMultiplicity = 0;
+                    for (final BaseEdge edge : outgoingEdges) {
+                        totalOutgoingMultiplicity += edge.getMultiplicity();
+                    }
+
+                    for (final E edge : outgoingEdges) {
+                        final V targetVertex = graph.getEdgeTarget(edge);
+                        queue.add(new KBestHaplotype<>(pathToExtend, edge, totalOutgoingMultiplicity));
+                    }
+                }
+            }
+        }
+        return result;
+    }
+
+    /**
+     * Removes edges that produces cycles and also dead vertices that do not lead to any sink vertex.
+     * @return never {@code null}.
+     */
+    protected BaseGraph<V, E> removeCyclesAndVerticesThatDontLeadToSinks(final BaseGraph<V, E> original, final Collection<V> sources, final Set<V> sinks) {
+        final Set<E> edgesToRemove = new HashSet<>(original.edgeSet().size());
+        final Set<V> vertexToRemove = new HashSet<>(original.vertexSet().size());
+
+        boolean foundSomePath = false;
+        for (final V source : sources) {
+            final Set<V> parentVertices = new HashSet<>(original.vertexSet().size());
+            foundSomePath = findGuiltyVerticesAndEdgesToRemoveCycles(original, source, sinks, edgesToRemove, vertexToRemove, parentVertices) || foundSomePath;
+        }
+
+        Utils.validate(foundSomePath, () -> "could not find any path from the source vertex to the sink vertex after removing cycles: "
+                    + Arrays.toString(sources.toArray()) + " => " + Arrays.toString(sinks.toArray()));
+
+        Utils.validate(!(edgesToRemove.isEmpty() && vertexToRemove.isEmpty()), "cannot find a way to remove the cycles");
+
+        final BaseGraph<V, E> result = original.clone();
+        result.removeAllEdges(edgesToRemove);
+        result.removeAllVertices(vertexToRemove);
+        return result;
+    }
+
+    /**
+     * Recursive call that looks for edges and vertices that need to be removed to get rid of cycles.
+     *
+     * @param graph the original graph.
+     * @param currentVertex current search vertex.
+     * @param sinks considered sink vertices.
+     * @param edgesToRemove collection  of edges that need to be removed in order to get rid of cycles.
+     * @param verticesToRemove collection of vertices that can be removed.
+     * @param parentVertices collection of vertices that preceded the {@code currentVertex}; i.e. the it can be
+     *                       reached from those vertices using edges existing in {@code graph}.
+     *
+     * @return {@code true} to indicate that the some sink vertex is reachable by {@code currentVertex},
+     *  {@code false} otherwise.
+     */
+    private boolean findGuiltyVerticesAndEdgesToRemoveCycles(final BaseGraph<V, E> graph,
```

Comment:
I agree. Indeed I would also comment that the vertexes that don't go to sinks can't actually occur at this stage in the code and would have been pruned before. Ultimately my answer to this would probably be to remove them entirely though since it doesn't matter in the new assembler....